// apollo-server-express/dist/expressApollo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const url_1 = __importDefault(require("url"));
const apollo_server_core_1 = require("apollo-server-core");
const GraphiQL = __importStar(require("apollo-server-module-graphiql"));
function graphqlExpress(options) {
    if (!options) {
        throw new Error('Apollo Server requires options.');
    }
    if (arguments.length > 1) {
        throw new Error(`Apollo Server expects exactly one argument, got ${arguments.length}`);
    }
    const graphqlHandler = (req, res, next) => {
        apollo_server_core_1.runHttpQuery([req, res], {
            method: req.method,
            options: options,
            query: req.method === 'POST' ? req.body : req.query,
            request: apollo_server_core_1.convertNodeHttpToRequest(req),
        }).then(gqlResponse => {
            res.setHeader('Content-Type', 'application/json');
            res.setHeader('Content-Length', Buffer.byteLength(gqlResponse, 'utf8').toString());
            res.write(gqlResponse);
            res.end();
        }, (error) => {
            if ('HttpQueryError' !== error.name) {
                return next(error);
            }
            if (error.headers) {
                Object.keys(error.headers).forEach(header => {
                    res.setHeader(header, error.headers[header]);
                });
            }
            res.statusCode = error.statusCode;
            res.write(error.message);
            res.end();
        });
    };
    return graphqlHandler;
}
exports.graphqlExpress = graphqlExpress;
function graphiqlExpress(options) {
    const graphiqlHandler = (req, res, next) => {
        const query = req.url && url_1.default.parse(req.url, true).query;
        GraphiQL.resolveGraphiQLString(query, options, req).then(graphiqlString => {
            res.setHeader('Content-Type', 'text/html');
            res.write(graphiqlString);
            res.end();
        }, error => next(error));
    };
    return graphiqlHandler;
}
exports.graphiqlExpress = graphiqlExpress;
